File: MSSR_2.0.0.py
import numpy as np
import math
import scipy.interpolate as interpolate
from napari.layers import Image, Labels, Layer, Points
import matplotlib.pyplot as plt

#Bicubic Interpolation
def bicInter(img, amp, mesh):
	width, height = img.shape
	y=np.linspace(1, width, width)
	x=np.linspace(1, height, height)
	imgInter=interpolate.interp2d(x, y, img, kind='cubic')
	y2=np.linspace(1, width, width*amp)
	x2=np.linspace(1, height, height*amp)
	Z2 = imgInter(x2, y2)
	if mesh:
		Z2 = meshing(Z2, amp)
	return Z2

#Fourier Interpolation
def ftInterp(img, amp, mesh):
	width, height = img.shape
	mdX = math.ceil(width/2) + 1
	mdY = math.ceil(height/2) + 1 
	
	extraBorder = math.ceil(amp/2)
	Nwidth = (width*amp) + extraBorder
	Nheight = (height*amp) + extraBorder
	
	lnX = len(np.arange((mdX),width))
	lnY = len(np.arange((mdY),height))
	
	imgFt = np.fft.fft2(img)
	imgFt = imgFt * (Nwidth/width) * (Nheight/height)
#	imgFt = imgFt * amp * amp
	
	fM = np.zeros((Nwidth, Nheight), dtype=complex)
	fM[0:mdX, 0:mdY] = imgFt[0:mdX, 0:mdY]; #izq sup cuadrante
	fM[0:mdX, (Nheight-lnY):Nheight] = imgFt[0:mdX, (mdY):height]; #der sup cuadrante
	fM[(Nwidth-lnX):Nwidth, 0:mdY] = imgFt[(mdX):width, 0:mdY]; #izq inf cuadrante
	fM[(Nwidth-lnX):Nwidth, (Nheight-lnY):Nheight] = imgFt[(mdX):width, (mdY):height]; #der inf cuadrante

	Z2 = (np.fft.ifft2(fM)).real
	Z2 = Z2[0:(width*amp), 0:(height*amp)]
	if mesh:
		Z2 = meshing(Z2, amp)
	return Z2

#Mesh compensation
def meshing(img, amp):
    width, height = img.shape
    desp = math.ceil(amp/2)
    imgPad = np.pad(img, desp, 'symmetric')
    imgS1 = imgPad[0:width, desp:height+desp]
    imgS2 = imgPad[(desp*2):width+(desp*2), desp:height+desp]
    imgS3 = imgPad[desp:width+desp, 0:height]
    imgS4 = imgPad[desp:width+desp, (desp*2):height+(desp*2)]
    imgF = (img + imgS1 + imgS2 + imgS3 + imgS4) / 5
    return imgF

#Spatial MSSR
def sfMSSR(img, fwhm, amp, order, mesh = True, ftI = False, intNorm = True):
	hs = round(0.5*fwhm*amp)
	if hs < 1:
		hs = 1
	maxValueImgOr = (max(map(max, img)))
	if(amp > 1 and not ftI):
		img = bicInter(img, amp, mesh)
	elif(amp > 1 and ftI):
		img = ftInterp(img, amp, mesh)
	width, height = img.shape
	xPad = np.pad(img, hs, 'symmetric')
	M = np.zeros((width,height))
	for i in range(-hs, hs+1):
		for j in range(-hs, hs+1):
			if i!=0 or j!=0:
				xThis = xPad[hs+i:width+hs+i, hs+j:height+hs+j]
				M = np.maximum(M, np.absolute(img-xThis))

	weightAccum = np.zeros((width,height))
	yAccum = np.zeros((width,height))

	for i in range(-hs, hs+1):
		for j in range(-hs, hs+1):
			if i!=0 or j!=0:
				spatialkernel = np.exp(-(pow(i,2)+pow(j,2))/pow((hs),2))
				xThis = xPad[hs+i:width+hs+i, hs+j:height+hs+j]
				xDiffSq0 = pow((img-xThis)/M,2)
				intensityKernel = np.exp(-xDiffSq0)

				weightThis = spatialkernel*intensityKernel
				weightAccum = weightAccum + weightThis
				yAccum = yAccum + (xThis*weightThis)

	MS = img - (yAccum/weightAccum)
	MS[MS < 0] = 0
	MS[np.isnan(MS)] = 0

	I3 = MS/(max(map(max, MS)))
	x3 = img/(max(map(max, img)))
	for i in range(order):
		I4 = x3 - I3
		I5 = max(map(max, I4)) - I4
		I5 = I5/max(map(max, I5))
		I6 = I5*I3
		I7 = I6/max(map(max, I6))
		x3 = I3
		I3 = I7
	I3[np.isnan(I3)] = 0
	if intNorm:
		IMSSR = I3*img
	else:
		IMSSR = I3
	return IMSSR

#Temporal MSSR
def tMSSR(img_layer, fwhm, amp, order, mesh = True, ftI = False, intNorm = True):
	img=np.array(img_layer.data)
	nFrames, width, height = img.shape
	imgMSSR = np.zeros((nFrames,width*amp,height*amp))
	for nI in range(nFrames):
		print("Image " + str(nI+1))
		imgMSSR[nI, :, :] = sfMSSR(img[nI], fwhm, amp, order, mesh, ftI, intNorm)
	return imgMSSR

#Mean
def tMean(img):
	return np.mean(img, 0)

#Variance
def tVar(img):
	return np.var(img, 0)

#Temporal Product Mean
def TPM(img):
    nFrames, width, height = img.shape
    SumTPM = np.zeros((width,height))
    iTPM = np.zeros((width,height))
    for i in range(nFrames):
        SumTPM = SumTPM + img[i]
    for i in range(nFrames):
        iTPM = iTPM + (SumTPM * img[i])
    return iTPM

#Auto-Cummulants
def TRAC(img, k):
    nFrames, width, height = img.shape
    avg = np.mean(img, 0)
    d0 = np.absolute(img - avg)
    d1 = d0[1:nFrames, :, :]
    d2 = d0[2:nFrames, :, :]
    d3 = d0[3:nFrames, :, :]
    if k == 2:
        trac = np.mean(d0[1:nFrames, :, :]*d1,0)
    elif k == 3:
        trac = np.mean(d0[2:nFrames, :, :]*d1[1:nFrames, :, :]*d2,0)
    else:
        t1 = np.mean(d0[3:nFrames, :, :]*d1[2:nFrames, :, :]*d2[1:nFrames, :, :]*d3,0)
        t2 = np.mean(d0[3:nFrames, :, :]*d1[2:nFrames, :, :],0)*np.mean(d2[1:nFrames, :, :]*d3,0)
        t3 = np.mean(d0[3:nFrames, :, :]*d2[1:nFrames, :, :],0)*np.mean(d1[2:nFrames, :, :]*d3,0)
        t4 = np.mean(d0[3:nFrames, :, :]*d3,0)*np.mean(d1[2:nFrames, :, :]*d2[1:nFrames, :, :],0)
        trac = np.absolute(t1-t2-t3-t4)
    return trac

#Empirical cumulative distribution function
def ecdf(data):
	""" Compute ECDF """
	data = np.reshape(data, -1)
	x = np.sort(data)
	n = len(x)
	y = np.arange(1, n+1) / n
	return(x,y)

#Exclude Outliers
def excOutliers(data, th):
	th = (100 - th)/100
	x, f = ecdf(data)
	found = np.where(f > th)
	mnX = x[found[0][0]];
	data = np.where(data >= mnX, mnX, data)
	return data

File: MSSR_2.0.0.R
library(FCSlib)
library(akima)
library(matlab)

# Stack analysis
tMSSR <- function(img, amp, psf, order, mesh = TRUE, intNorm = TRUE){#, excOL = FALSE, OutLiersTh = 0.3){
  di <- dim(img)
  MSSRDFinal <- array(data = NA, dim = c(di[1]*amp, di[2]*amp, di[3]))
  for(i in 1:dim(img)[3]){
    MSSRDFinal[,,i] <- sfMSSR(img = img[, , i], amp = amp, psf = psf, order = order, mesh = mesh,
                              intNorm = intNorm)
  }
  return(MSSRDFinal)
}
# Single frame analysis
sfMSSR <- function(img, amp, psf, order, mesh = TRUE, intNorm = TRUE) {
  di <- dim(img)
  hs <- round(0.5*psf*amp)
  if(hs<1){
    hs <- 1
  }
  if(amp > 1){
    AMP <- bicubic.grid(x = 1:di[1], y = 1:di[2], z = img, nx = di[1]*amp, ny = di[2]*amp)$z
    if(mesh){
      AMP <- meshing(img = AMP, desp = ceiling(amp/2))
    }
  } else{
    AMP <- img
  }
  xPad <- padarray(AMP, c(hs, hs), "symmetric", "both")
  
  M <- array(data = 0, dim = dim(AMP))
  di <- dim(M)
  for(i in -hs:hs){
    for(j in -hs:hs){
      if((i!=0 || j!=0) && (sqrt(i^2 + j^2) <= hs)){
        xThis <- xPad[(hs+i+1):(di[1]+hs+i), (hs+j+1):(di[2]+hs+j)]
        M = pmax(M, abs(AMP - xThis))
      }
    }
  }
  
  weightAccum <- yAccum <- array(data = 0, dim = dim(AMP))
  for(i in -hs:hs){
    for(j in -hs:hs){
      if((i!=0 || j!=0) && (sqrt(i^2 + j^2) <= hs)){
        spatialKernel <- exp(-(i^2 + j^2) / (hs^2))
        xThis <- xPad[(hs+i+1):(di[1]+hs+i), (hs+j+1):(di[2]+hs+j)]
        xDiffSq0 <- ((AMP - xThis) / M)^2
        intensityKernel <- exp(-xDiffSq0)
        
        weightThis <- spatialKernel * intensityKernel
        weightAccum <- weightAccum + weightThis
        yAccum <- yAccum + (xThis * weightThis)
      }
    }
  }
  
  MS <- AMP - (yAccum / weightAccum)
  MS[which(MS<0)] <- 0
  MS[which(is.na(MS))] <- 0
  
  I0 <- MS / max(MS)
  I3 <- I0
  x3 <- AMP / max(AMP)
  if(order>0){
    for(i in 1:order){
      I4 <- x3 - I3
      I5 <- max(I4) - I4
      I5 <- I5 / max(I5)
      I6 <- I5 * I3
      I7 <- I6 / max(I6)
      x3 <- I3
      I3 <- I7
    }
  }
  IMSSRD <- rot90(I3,3)
  if(intNorm){
    IMSSRD <- IMSSRD*AMP# Intensity Normalization
  }
  return(IMSSRD)
}
excludeOutliers <- function(IMSSRD, OutLiersTh){
  th = (100-OutLiersTh)/100
  
  IMSSRD <- IMSSRD*max(AMP)
  # print(min(IMSSRD))
  # print(max(IMSSRD))
  x <- seq(min(IMSSRD),max(IMSSRD), 1)
  Fn <- ecdf(IMSSRD)
  f <- Fn(x)
  x[which(f>th)]
  mnX <- min(x[which(f>th)])
  IMSSRD[which(IMSSRD>mnX)] <- mnX
  IMSSRD <- IMSSRD/max(AMP)
  
  x <- seq(min(AMP),max(AMP), 1)
  Fn <- ecdf(AMP)
  f <- Fn(x)
  x[which(f>th)]
  mnX <- min(x[which(f>th)])
  AMP[which(AMP>mnX)] <- mnX
}
# Mesh Compensation
meshing <- function(img, desp){
  img <- padarray(img, c(desp, desp), "symmetric", "both") # Genera una matriz reflejada en todas las dimensiones
  di <- dim(img)
  imgHD <- rbind(img[(di[1]-desp):di[1],], img[1:(di[1]-desp-1),])
  imgHI <- rbind(img[(desp+1):di[1],], img[1:(desp),])
  
  imgVD <- cbind(img[, (di[2]-desp):di[2]], img[, 1:(di[2]-desp-1)])
  imgVI <- cbind(img[, (desp+1):di[2]], img[, 1:(desp)])
  
  imgHVC <- (img + (imgHD)  + (imgHI) + (imgVD)  + (imgVI)) / (5)
  return(imgHVC[(desp+1):(di[1]-desp), (desp+1):(di[2]-desp)])
}
# Temporal pixel-wise analysis
TPM <- function(img){
  di <- dim(img)
  SumTPM <- array(data = 0, dim = di[1:2])
  iTPM <- array(data = 0, dim = di[1:2])
  for(i in 1:di[3]){
    SumTPM <- SumTPM + img[,,i]
  }
  for(i in 1:di[3]){
    iTPM <- iTPM + (img[,,i] * SumTPM)
  }
  return(iTPM)
}
imVar<-function(img){
  di<-dim(img)
  X<-di[1]
  Y<-di[2]
  Ivar<-array(NA, dim = c(X,Y))
  for(i in 1:(X)){ 
    for(j in 1:(Y)){  
      Ivar[i,j]<-var(img[i,j,], na.rm = T)
    }
  }
  return(Ivar)
}
imMean<-function(img){
  di<-dim(img)
  if(length(di)!=3) return(img)
  if(di[3]<2) return(img)
  X<-di[1]
  Y<-di[2]
  Imean<-array(NA, dim = c(X,Y))
  for(i in 1:(X)){ 
    for(j in 1:(Y)){  
      Imean[i,j]<-mean(img[i,j,], na.rm = T)
    }
  }
  return(Imean)
}